code-ready/crc
# [WIP] Restructuring tests: an example of splitting basic tests into integration and e2e

## Idea

This PR introduces initial changes so we have a concrete example when we talk about restructuring the CRC tests.

By adhering to Cucumber's design principles (thus not needing BeforeFeature hooks), we can upgrade to godog v0.11.0 and get support for jUnit file export. This means that we need to shift our more detailed (integration) tests away from Cucumber. Ginkgo was suggested as a possible candidate.

So this is a preview of splitting the tests into _integration_ and _e2e_ on the example of `basic` tests. The integration part is in Ginkgo and the e2e part is in Cucumber.

I reused a lot of Guillaume's Ginkgo set-up code from Tea Time discussion.

**Still missing**
- Test on all 3 platforms (only Linux has been tested)
- Test with standalone bundle (only release bits were tested)
- Various combinations of small commands in Ginkgo tests. See [this map](https://drive.google.com/file/d/1Zdk5zm-PLkiDJTFaGeJxVfs_zaCp64jW/view?usp=sharing). The `config`, `ip`, `console`, `oc-env`, `podman-env` subtrees are not tested yet. Many flag combinations are also missing as of now.
- Need to designate a _focus_ subset of basic tests to run as an "important" subset when not everything needs to be tested.
- Did not update dependency to godog v0.11.0, hence no jUnit output for e2e tests yet.

**Warning**
- Renamed `PULL_SECRET_FILE` env var to `PULL_SECRET_LOCATION` to be analogous to `BUNDLE_LOCATION` var. This `Makefile` change will affect pipelines.

## Proposed changes

1. Move the integration component of the `basic` tests to Ginkgo framework in `test/ginkgo`
2. Add disk/mem/cpu resize integration tests in Ginkgo.
3. Condense the Cucumber e2e tests component to adhere to Cucumber design principles. E.g. independent Scenarios.
4. Produce jUnit files with test results to upload to ReportPortal (or elsewhere). 

## Testing

1. Set `BUNDLE_LOCATION`(if using standalone bundle) and `PULL_SECRET_LOCATION` variables.
2. Go to `test/ginkgo` and trigger the integration tests by typing `ginkgo`. Check jUnit output in `test/ginkgo/results`.
3. To run e2e basic test, modify makefile to only run `<os> && basic` tags and type `make integration` from repo dir.
